### PR TITLE
Cranelift update

### PR DESCRIPTION
**Pull request by xdoardo**
An in-progress effort to upgrade cranelift.

Side quests: 
+ [x] update the MSRV to 1.78. 
+ [x] update `wat` to latest
+ [x] update `wasmparser`

### CODE REVIEW

File: lib/compiler-cranelift/src/config.rs [hunk 1]
```diff
@@ -137,11 +136,11 @@ impl Cranelift {
...
-        // There are two possible traps for division, and this way
-        // we get the proper one if code traps.
-        flags
-            .enable("avoid_div_traps")
-            .expect("should be valid flag");
+        //// There are two possible traps for division, and this way
+        //// we get the proper one if code traps.
+        //flags
+        //    .enable("avoid_div_traps")
+        //    .expect("should be valid flag");
```
Comment (syrusakbary) [inline lib/compiler-cranelift/src/config.rs:143]: This setting no longer exists in cranelift? If so, can we just remove this?

File: lib/compiler-cranelift/src/config.rs [hunk 2]
```diff
@@ -177,15 +176,15 @@ impl Cranelift {
...
-        }
+        //if is_riscv {
+        //    flags
+        //        .set("enable_simd", "false")
+        //        .expect("should be valid flag");
+        //} else {
+        //    flags
+        //        .set("enable_simd", "true")
+        //        .expect("should be valid flag");
+        //}
```
Comment (syrusakbary) [inline lib/compiler-cranelift/src/config.rs:187]: Remove me

File: lib/api/src/sys/tunables.rs [hunk 3]
```diff
@@ -255,6 +255,7 @@ mod tests {
     }
 
     #[test]
+    #[cfg(feature = "cranelift")]
```
Comment (syrusakbary) [inline lib/api/src/sys/tunables.rs:258]: This should be for any engine. Undo

File: lib/compiler-cranelift/Cargo.toml [hunk 4]
```diff
@@ -14,21 +14,32 @@ rust-version.workspace = true
...
+	"isle-in-source-tree"
+] }
+cranelift-frontend = { version = "0.110.2", default-features = false }
+itertools = "0.12.0"
 tracing = "0.1"
 hashbrown = { version = "0.11", optional = true }
 rayon = { version = "1.5", optional = true }
 more-asserts = "0.2"
-gimli = { version = "0.26", optional = true }
+gimli = { version = "0.28.0", optional = true }
```
Comment (syrusakbary) [inline lib/compiler-cranelift/Cargo.toml:37]: Nice. can we update gimli as well on the singlepass crate and make the version global for the Cargo workspace?

File: lib/compiler-cranelift/src/func_environ.rs [hunk 5]
```diff
@@ -119,7 +125,9 @@ pub struct FuncEnvironment<'module_environment> {
     /// The memory styles
     memory_styles: &'module_environment PrimaryMap<MemoryIndex, MemoryStyle>,
 
-    /// The table styles
+    /// Cranelift tables we have created to implement Wasm tables.
+    tables: SecondaryMap<TableIndex, Option<TableData>>,
```
Comment (syrusakbary) [inline lib/compiler-cranelift/src/func_environ.rs:129]: Why this is an option?

It seems to me it should be:

```suggestion
    tables: SecondaryMap<TableIndex, TableData>,
```

File: lib/compiler-cranelift/Cargo.toml (hunk 4 above)
Comment (syrusakbary) [reply lib/compiler-cranelift/Cargo.toml:37]: Note: use the same gimli version as `cranelift-codegen` (that is: `0.28` (.1) )

File: lib/compiler-cranelift/src/func_environ.rs (hunk 5 above)
Comment (xdoardo) [reply lib/compiler-cranelift/src/func_environ.rs:129]: The reason, I think, is that there's no sensible default value for `TableData`, and creating a new `SecondaryMap` (both from `new` or `default`) [requires the value type to implement `Default`](https://docs.rs/cranelift-entity/latest/cranelift_entity/struct.SecondaryMap.html#method.new). Perhaps, not using `SecondaryMap`s might be viable as well.

**Review by syrusakbary — APPROVED** [review]

**Review by syrusakbary — APPROVED** [review]

**Pull request merged**